### ISSUE TITLE
PDF / Document component (Carbone Integration?)

### ISSUE THREAD

**Issue by joebudi** [Budibase/budibase#1372]
As a land surveyor, I would like to generate pdfs from table data, so I can store and print them (possibly to send to clients).

More info can be found here:
#1353

**Comment by shogunpurple** [comment]
Nope - very relevant

**Comment by Bogey** [comment]
I would welcome this too. Preferably a pre-created template (either DOCX, ODT or PDF) with handlebar variables that are then used to create the PDF.

**Comment by Manishkparmar80** [comment]
It is required. Pls refer post ideas # 1696

**Comment by Manishkparmar80** [comment]
@joebudi is this feature will be avaible in July 21 release ?

**Comment by joebudi** [comment]
Hey @Manishkparmar80 

@aptkingston had a stab at a report / PDF component during the company's lab day last week. There's still a lot more work to do to complete the component. It's not on our roadmap for this month. But, there does seem to be demand for it so I might be able to push for it's completion. 

To help me get the work approved, I will need you to detail, very clearly:
- why you need the component (use case)
- how you expect it to work
- what you expect it to do

_P.s this comment is not confirmation that the component will be delivered for the July release. _

**Comment by Bogey** [comment]
**why you need the component (use case)**
A PDF is the new paper (the next best thing) because of it's psychological effect. A report in PDF adds some sort professional weight above an email. It offers to be a document in line with corporate identity, and if printed it maintains that identity. The reports I create are summaries of audits done, numbers analysed and displayed as graphs, and completed with a final recommendation.

**how you expect it to work**
To take out the guesswork of how the PDF will look like the template is designed as DOCX, ODT or PDF.
Conversion from HTML proofs still unreliable with the positioning of a/o images.
Templates are created upfront with either handlebar, square brackets or other declaration form for variables.
Example:
---
_Header of report_.
_Prepared for: {{variable1}}_
Aenean eu leo quam. Pellentesque ornare sem lacinia quam venenatis vestibulum.

_Subheader_
Current situation can be described as: {{variable2}}

_Recommendation_
Our recommendation for {{variable1}} is to {{variable3}}

 {{variable4}}
 {{imagevariable1}}
---
Via a special prep-form I select with what and how the variables are filled.
variable1 becomes client X's name that is selected from a dropdown select.
Picking client X from the dropdown selects variable2 to be automatically filled with data stored and, of course, linked to client X.
Variable3 has some logic: if clientX.profit > 250k then variable3 becomes recommendation ABC. 
if clientX.profit < 250k then variable3 becomes recommendation XYZ.

Then the data is pushed to create a PDF. This PDF can then be send via email, stored for later access.

**what you expect it to do**
Preferably PDFs can be digitally signed and/or encrypted.
Also permissions can be set (can user print, copy, extract or delete pages yes/no).

I hope this helps?
Cheers.

**Comment by sergey-dme** [comment]
As for the use case, I'd like to to add the following:

Whereas Budibase is being used as an internal tool to automate certain processes in organizations, there are often QMS/CMS requirements to sign the final document or form (soft or hard copy) and then store/save it. The PDF is most suitable and usable format for it.

**Comment by Manishkparmar80** [comment]
Other use case : 

E.g. 

1. In store we book the order for item then they need to share Invoice and Delivery challan with customer. Invoice , delivery challan etc are report.

2. If you visit any grocery shop then after payment they give us printed bill which is also report.

3.After online shopping,  we receive purchase bill or invoice in PDF format by email in digitally signed format. 

Reports are heart of any application or software. Without report, application can be think of useful. 

Application must have provision to provide report in multiple form PDF, DOC , excel etc based on business requirement.

**Comment by Manishkparmar80** [comment]
@joebudi 
Yesterday come across few more  real time use cases :

1. Making online transaction via bank and it provide me option of downloading transaction receipt in PDF , Word or excel.

2. While downloading my bank statement, website providing me option of downloading  bank statement by PDF , excel, work or CSV.

**Comment by Manishkparmar80** [comment]
@joebudi Hi, is this feature/ functionality( PDF reporting)  is considered for product enhancement?

**Comment by joebudi** [comment]
Hey @Manishkparmar80 

We've talked about this internally and we will need PDF support in the near future. The extent of the functionality is to be decided. There are no plans to complete this work, this month. We have an IPM meeting at the end of the month, where I will bring this discussion forward, to the team. We will then make a decision on whether or not it is to be completed in the next iteration.

During a lab day (2 weeks back), @aptkingston made decent progress on his interpretation of what this would look like. So, there is 1 version of PDF support nearly* complete, but it does not contain some functionality requested in this discussion, nor does it follow the same UX.

* I am not sure how complete this is.

**Comment by Manishkparmar80** [comment]
@joebudi  Thanks for sharing update.

**Comment by shogunpurple** [comment]
bump

**Comment by mjashanks** [comment]
Just saying that I had 2 more people say to me today on calls that "Printing Documents" is one feature that every other low/no-code platform is missing.

**Comment by pepeday** [comment]
Bubble has this feature (more or less)

**Comment by igbojionu** [comment]
wow whats the update on this an patiently waiting

**Comment by kasgithub8** [comment]
@igbojionu It's on the Roadmap for Q3 this Year!

**Comment by neoguet** [comment]
Hi
I wrote here my needs for a PDF integration in BB.

I would like to be able to generate a PDF file when a row is created or updated via an action or automation and after this file is saved in a field of the row.
Then I would like to be able to send this file via email.
the construction of this file could be done with a template and handlebars syntax as for onboarding mails.
[jsPDF](https://github.com/parallax/jsPDF) is maybe a solution?

**Comment by oswaldB** [comment]
Hello if I may,

I would also suggest to use the jsreport to create pdf rather than Carbone.

**Comment by berief** [comment]
Hello there,

is there any update when the PDF funcionality will be published? Thank you.

**Comment by lburtwistle** [comment]
@neoguet - I just heard the same requirements today form a client. Being able to generate and download a pdf with official header would take it to the next level from an enterprise prospective.  Right now taking the client to a screen to review costs, so would be great to download a pdf version of it when doing the review.  Also, having the ability to include the pdf in automation, so you could send it out within a notification when the review is completed.  It really has so many use cases when having to share an official document with outside entities - invoicing, purchase orders, billing, survey, bill of lading, etc.

**Comment by mahxds** [comment]
Yep, Companies that use Low-Code to get fast and easy app fitting their unique workflow mostly are that small or have overlapping situations with external service providers and other stuff , who need need paperwork.

Thats why it would be very great to get a small and simple component that will put a PDF printout with value on predefined "positions". 

Like, define bindings and their positions on the PDF, upload a PDF Template. And done.
Add this as action, so you get generated PDF with the values which you can print.

**Comment by Voroxpete** [comment]
We're looking at a number of potential use cases for Budibase at my company and this is considered a make-or-break for many of them.  Just throwing my hat in the ring for "This is a really important feature" I guess.

**Comment by davidpfahler** [comment]
I found this issue trying to build an invoice generator app. Seems like a basic feature lots of people need. What would be the best way to contribute to this feature?

**Comment by fdecourt** [comment]
Agreed with all the others, this is mandatory.....

**Comment by khalid-al-walid** [comment]
any updates about this pdf generator?

**Comment by lburtwistle** [comment]
Yes, I too am anxiously waiting any updates on prioritizing this one :)

**Comment by mahxds** [comment]
same would be so awesome to be able to easly geneare pdf

**Comment by taholmes160** [comment]
why you need the component (use case) 
My company moves cars between sale sites -- being able to hand the drivers a list of cars to pick up is huge, as well -- passing data on about what has been moved to the operations team would be very helpful

how you expect it to work 
Select the information needed (filter or query) and dump it through a template to output as a .csv, .pdf etc

what you expect it to do 
Allow me to create reports

**Comment by acseven** [comment]
https://github.com/Budibase/budibase/issues/1372#issuecomment-860061523
> Nope - very relevant

Still very relevant!

**Comment by Cloud-Kid** [comment]
Hey, still a very interesting feature, it's indeed really needed.

The ability to generate PDF simply using form data or/and the ability to generate a PDF using one line of form at a time to complete PDF fields for instance.

What can be added later is notifications support during the PDF creation workflow.

Big thanks for your work !

**Comment by rawtaz** [comment]
Can we get the bot to not be an idiot and wanting to close this all the time? :-) This feature is a super important one.

**Comment by aNobre-redhat** [comment]
Moved from Q3 / 22 to Q3 / 23 roadmap? Such a shame. This feature is a must have.

**Comment by AlejandroFNadal** [comment]
We ended up making a Python api endpoint to generate the PDF's, and then downloading them as base64 over the api and regenerating them as pdf files using javascript in a dummy "Set state". We hope that we will be able to migrate this feature to the future budibase pdf feature.

**Comment by gwlewis** [comment]
This is a very important feature for our company as well.

**Comment by Mcc83** [comment]
This is a must have. My company prefers to have a paper record of everything. Budibase needs some more exporting/pdf/print screen tools for sure. Would love to see this go further, hopefully in 2024.

**Comment by Bewo** [comment]
I recently started using Budibase and I have to say it's a really great project!
I'm looking forward to working with it! But when I realised in the middle of my first project that there is no way to export PDFs natively, I was shocked! I can only agree with the previous speakers: In my opinion too, this is an absolute MUST-HAVE for such a great project!

Just a thought: If it's too much effort to integrate Carbone, then perhaps a "simple interface" like [mpdf ](https://github.com/mpdf/mpdf)would suffice. This way every user has the possibility to define his content using custom CSS and Html/JS...

**Comment by leonlazic** [comment]
Please add the PDF generation function to the Q1 of 2024.

**Comment by leonlazic** [comment]
👀 The new global bindings demo also shows a little of the upcoming PDF generator component ...

[Global Bindings Demo video](https://player.vimeo.com/video/908298966?autoplay=1&muted=1&loop=1&autopause=0&controls=1&autohide=1;badge=0;amp;quality_selector=1&player_id=0&app_id=58479)

**Comment by tariktokic** [comment]
Super important, must have feature! Actually a deal breaker for us in the company. We need to have a PDF generated after submitting a form, so it can be printed and signed by the customer.

**Comment by dthompsonusa** [comment]
I'm evaluating Budibase.  The application being built requires the ability for the user to press a button that then runs code to generate a set of reports from a combination of related tables.  From this thread it appears that Budibase is not capable of doing this.  Please confirm if it is or not and if not when that functionality will be available.  Thank you.

**Comment by mjashanks** [comment]
@dthompsonusa Budibase does not have an in-built PDF generator. However, it is capable of file handling in automations. This means that you can use a 3rd party report generator (e.g Carbone) to, Save, Email or return PDFs to the end-user in-app.

There's a great guide here: https://budibase.com/blog/updates/2024/files-in-automation/

**Comment by dthompsonusa** [comment]
Thank you Michael for the info.  I'm trying to follow the guide but am getting stuck on where to put the json 
{
	"data":{{Binding.data}},
  "convertTo": "pdf",
  "currencySource" : "GBP",
  "currencyTarget" : "GBP"
}

The interface in the guide seems to be different than the current version of budibase.  Any help is greatly appreciated.

**Comment by JBSDS** [comment]
@dthompsonusa Have you found a solution? I have the same problem.

**Comment by joebudi** [comment]
This should help: 

https://budibase.com/blog/updates/2024/files-in-automation/

**Comment by melohagan** [comment]
https://docs.budibase.com/docs/pdf-exporting

**Comment by rawtaz** [comment]
Very cool, I haven't tried it yet but it looks nice reading the docs! May I ask, which PRs or code changes was this implemented in?

**Comment by melohagan** [comment]
Hey @rawtaz It was added three months ago! https://docs.budibase.com/changelog/dynamic-pdfs

The GitHub issue had just been missed. 

The release version was [3.8.4](https://github.com/Budibase/budibase/releases/tag/3.8.4)
